### PR TITLE
feat: impl fundamental signature verification modes

### PR DESCRIPTION
**Pull request by EthanYuan**
Based on dynamic linking, the following 3 modes of otx locking scripts are implemented

1 compatible mode: fully compatible with `secp256k1_blake2b_sighash_all`
2 sighash mode: `Single | AnyoneCanPay`
3 sighash mode: `ALL | AnyoneCanPay`

### CODE REVIEW

**Comment by doitian** [comment]
Let's call it otx-sighash-lock to differentiate it with the commands-list version.

**Comment by EthanYuan** [comment]
> Let's call it otx-sighash-lock to differentiate it with the commands-list version.

Done

File: contracts/otx-lock/src/helper.rs [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+
+    // compare with expected pubkey_hash
+    if &pubkey_hash[..] != expected_pubkey_hash {
+        return Err(Error::WrongPubkey);
+    }
+    Ok(())
+}
+
+pub(crate) fn add_prefix(sighash: u8, message: &mut [u8]) {
+    let mut prefix = Vec::new();
```
Comment (doitian) [inline contracts/otx-lock/src/helper.rs]: It's simpler to repeatedly call `blake2b.update`.

File: contracts/otx-lock/src/validate/sighash_all_acp.rs [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+    // input
+    let input = tx.inputs().get(index).ok_or(Error::Encoding)?;
+    let input_len = input.as_slice().len() as u64;
+
+    // outputs
+    let outputs = tx.outputs();
+    let outputs_count = outputs.len();
+    let outputs_len = outputs.as_slice().len() as u64;
+
+    // witness
```
Comment (doitian) [inline contracts/otx-lock/src/validate/sighash_all_acp.rs]: Leave a comment here: Sighash mode does not cover witnesses at positions beyond the number of inputs.

**Review by doitian — APPROVED** [review]
Posted two minor suggestions.

**Pull request merged**